servers/src/main/java/tachyon/worker/block/TieredBlockStore.java

```diff
@@ -342,104 +368,208 @@ public void registerBlockStoreEventListener(BlockStoreEventListener listener) {
+      long ownerUserId = tempBlockMeta.getUserId();
+      if (ownerUserId != userId) {
+        throw new InvalidStateException("ownerUserId of blockId " + blockId + " is " + ownerUserId
+            + " but userId passed in is " + userId);
+      }
+      return tempBlockMeta;
+    } finally {
+      mMetadataLock.readLock().unlock();
     }
-
-    return tempBlockMeta;
   }
 
-  // Abort a temp block. This method requires eviction lock in READ mode.
-  private void abortBlockNoLock(long userId, long blockId) throws NotFoundException,
+  // Abort a temp block.
+  private void abortBlockInternal(long userId, long blockId) throws NotFoundException,
       AlreadyExistsException, InvalidStateException, IOException {
-    TempBlockMeta tempBlockMeta = validateUserTempBlock(userId, blockId);
-    String path = tempBlockMeta.getPath();
-    if (!new File(path).delete()) {
-      throw new IOException("Failed to abort temp block " + blockId + ": cannot delete " + path);
+    long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+    try {
+      String path;
+      mMetadataLock.writeLock().lock();
+      try {
+        TempBlockMeta tempBlockMeta = checkTempBlockExists(userId, blockId);
+        path = tempBlockMeta.getPath();
+        mMetaManager.abortTempBlockMeta(tempBlockMeta);
+      } finally {
+        mMetadataLock.writeLock().unlock();
+      }
+
+      FileUtils.delete(new File(path));
+    } finally {
+      mLockManager.unlockBlock(lockId);
     }
-    mMetaManager.abortTempBlockMeta(tempBlockMeta);
+
   }
 
-  // Commit a temp block. This method requires eviction lock in READ mode.
-  private void commitBlockNoLock(long userId, long blockId, TempBlockMeta tempBlockMeta)
+  // Commit a temp block.
+  private BlockStoreLocation commitBlockInternal(long userId, long blockId)
       throws AlreadyExistsException, InvalidStateException, NotFoundException, IOException,
       OutOfSpaceException {
-    validateUserTempBlock(userId, blockId);
-    String sourcePath = tempBlockMeta.getPath();
-    String destPath = tempBlockMeta.getCommitPath();
-    FileUtils.move(new File(sourcePath), new File(destPath));
-    mMetaManager.commitTempBlockMeta(tempBlockMeta);
-  }
+    long lockId = mLockManager.lockBlock(userId, blockId, BlockLockType.WRITE);
+    try {
+      // When committing TempBlockMeta, its BlockMeta calculates the block size according to the
+      // size of the file of this TempBlockMeta. Therefore, commitTempBlockMeta must complete
+      // before moving actual block file to its committed path.
+      BlockStoreLocation loc;
+      String srcPath;
+      String dstPath;
+      mMetadataLock.writeLock().lock();
+      try {
+        TempBlockMeta tempBlockMeta = checkTempBlockExists(userId, blockId);
+        srcPath = tempBlockMeta.getPath();
+        dstPath = tempBlockMeta.getCommitPath();
+        loc = tempBlockMeta.getBlockLocation();
+        mMetaManager.commitTempBlockMeta(tempBlockMeta);
```

Comment:
There is a gap in time before the commit of the temp block, and the move to the committed path, so something could happen between the two operations. Will this be fine?

I want to say "yes", since the block lock is (usually) acquired before the metadata lock? What about the scenarios where the block lock is NOT acquired at all (but the metadata lock is)?